sentry-android-core/src/main/java/io/sentry/android/core/AndroidUserCache.java

```diff
@@ -0,0 +1,90 @@
+package io.sentry.android.core;
+
+import android.content.SharedPreferences;
+import io.sentry.core.IUserCache;
+import io.sentry.core.protocol.User;
+import io.sentry.core.util.Objects;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+// TODO: to think, this could be a SentryOptionsCache that could be expanded, does it make sense?
+
+/** the Android User Cache class. It caches the user data in the Android SharedPreferences */
+public final class AndroidUserCache implements IUserCache {
+
+  // TODO: only these 3 fields make sense to me, otherwise caching a raw json would be better, but I
+  // don't want to do that.
+  private static final String EMAIL = "email";
+  private static final String ID = "id";
+  private static final String USERNAME = "user_name";
+
+  private final @NotNull SharedPreferences sharedPreferences;
+  private final @NotNull SentryAndroidOptions options;
+  private final @Nullable String sentryDeviceId;
+
+  public AndroidUserCache(
+      final @NotNull SentryAndroidOptions options,
+      final @NotNull SharedPreferences sharedPreferences,
+      final @Nullable String sentryDeviceId) {
+    this.options = Objects.requireNonNull(options, "SentryAndroidOptions is required.");
+    this.sharedPreferences =
+        Objects.requireNonNull(sharedPreferences, "SharedPreferences is required.");
+    this.sentryDeviceId = sentryDeviceId;
+  }
+
+  //  TODO: I believe this makes sense only for global hub SDKs, but if not,
+  //   how do we cache multiple users (because we have multiple scopes)?
+
+  @Override
+  public void setUser(final @Nullable User user) {
+    if (options.isCacheUserForSessions()) {
+      final SharedPreferences.Editor edit = sharedPreferences.edit();
+      if (user != null) {
+        edit.putString(ID, user.getId())
+            .putString(EMAIL, user.getEmail())
+            .putString(USERNAME, user.getUsername())
```

Comment:
For me it is a bit confusing why we pass the whole user to `IUserCache.setUser` and only save id, email and username. Why don't we save the other users properties?

Reply:
Maybe then we should create an extra class for this and name it something like `SessionIdCache` and add this as a comment?

Reply:
I believe the Caching class should not be aware of what `id` it is, by that I mean (id, email or username).
maybe this class should take a String nominated as `distinctId` and that's all, so the naming would not `matter`.
A session also has an `id` so `SessionIdCache` would not fit, as we are caching the `Session.distinctId` which is a unique identifier of the user or device.
`UserIdCache` would make sense though?! naming is hard hehe

Reply:
> maybe this class should take a String nominated as distinctId

I prefer this too. It says what it does. So when setting the user, if this option is enabled, we select the data from the user object (id, username, email) to become the distinctId and cache that so the caching itself isn't aware of this logic and we don't have to deal with caching the actual user object and possible user-defined fields.